File: java/src/main/java/io/github/messageflow/CancellationTokenSource.java
package io.github.messageflow;

import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Creates {@link CancellationToken} instances and signals their cancellation.
 */
public final class CancellationTokenSource {

    private final AtomicBoolean cancelled = new AtomicBoolean();
    private final CancellationToken token = new CancellationToken(cancelled);

    /**
     * Gets the token signalled by this source.
     *
     * @return the token of this source
     */
    public CancellationToken token() {
        return token;
    }

    /** Signals cancellation to every holder of {@link #token()}. */
    public void cancel() {
        cancelled.set(true);
    }
}

File: node/src/CancellationTokenSource.ts
import { CancellationToken } from './CancellationToken.js';

/**
 * Creates {@link CancellationToken} instances and signals their cancellation.
 */
export class CancellationTokenSource {
  private readonly cancelled = { value: false };
  private readonly _token: CancellationToken;

  constructor() {
    this._token = CancellationToken['createFromSource'](this.cancelled);
  }

  /**
   * Gets the token signalled by this source.
   *
   * @returns the token of this source
   */
  token(): CancellationToken {
    return this._token;
  }

  /**
   * Signals cancellation to every holder of {@link token}.
   */
  cancel(): void {
    this.cancelled.value = true;
  }
}
